ambry-protocol/src/main/java/com/github/ambry/protocol/AmbryRequests.java

```diff
@@ -683,7 +716,154 @@ public void handleReplicaMetadataRequest(NetworkRequest request) throws IOExcept
+        replicateBlobRequest.getClientId(), MessageFormatFlags.All, partitionRequestInfoList,
+        GetOption.Include_All);
+    GetResponse getResponse = null;
+    try {
+      ConnectedChannel connectedChannel = connectionPool.checkOutConnection(remoteHostName, remoteHostPort, connectionPoolCheckoutTimeoutMs);;
+      ChannelOutput channelOutput = connectedChannel.sendAndReceive(getRequest);
+      getResponse = GetResponse.readFrom(channelOutput.getInputStream(), clusterMap);
+      if ((getResponse.getError() != ServerErrorCode.No_Error) || (getResponse.getPartitionResponseInfoList() == null)
+          || (getResponse.getPartitionResponseInfoList().size() == 0)) {
+        logger.error("ReplicateBlobRequest failed to get blob {} from the remote node {} {} {}",
+            blobId, remoteHostName, remoteHostPort, getResponse.getError());
+        response = new ReplicateBlobResponse(replicateBlobRequest.getCorrelationId(), replicateBlobRequest.getClientId(),
+            ServerErrorCode.Unknown_Error);
+      }
+    } catch (Exception e) {
+      logger.error("ReplicateBlobRequest getBlob {} from the remote node {} {} hit exception ",
+          blobId, remoteHostName, remoteHostPort, e);
+      response = new ReplicateBlobResponse(replicateBlobRequest.getCorrelationId(), replicateBlobRequest.getClientId(),
+          ServerErrorCode.Unknown_Error);
+    } finally {
+      if (response != null) {
+        long processingTime = SystemTime.getInstance().milliseconds() - startTime;
+        totalTimeSpent += processingTime;
+        publicAccessLogger.info("{} {} processingTime {}", replicateBlobRequest, response, processingTime);
+        metrics.replicateBlobProcessingTimeInMs.update(processingTime);
+        requestResponseChannel.sendResponse(response, request,
+            new ServerNetworkResponseMetrics(metrics.replicateBlobResponseQueueTimeInMs, metrics.replicateBlobSendTimeInMs,
+                metrics.replicateBlobTotalTimeInMs, null, null, totalTimeSpent));
+        return;
+      }
+    }
+
+    // step 2: get the blob message from the response and write to the store
+    // only have one partition. And checked at least it has one entry above.
+    PartitionResponseInfo partitionResponseInfo = getResponse.getPartitionResponseInfoList().get(0);
+    if (partitionResponseInfo.getErrorCode() == ServerErrorCode.No_Error) {
+      try {
+        List<MessageInfo> messageInfoList = partitionResponseInfo.getMessageInfoList();
+        if (messageInfoList == null || messageInfoList.size() != 1) {
+          logger.error("ReplicateBlobRequest PartitionResponseInfo response from GetRequest {} {} {} {} {} returned null.",
+              partitionResponseInfo, messageInfoList, remoteHostName, remoteHostPort, blobId);
+          response = new ReplicateBlobResponse(replicateBlobRequest.getCorrelationId(), replicateBlobRequest.getClientId(),
+              ServerErrorCode.Blob_Not_Found);
```

Comment:
Just a Q. Should we return internal error in this case since target server would actually return a response with valid error code when blob is not found.

Reply:
I assume the GetRequest may have a couple of partition/blob IDs to get.
Do you mean if there is a single BlobID is missing, the remote server will return a error code?

I see. Returning UNKNOWN_ERROR.